lib/recursive-methods.rb

```diff
@@ -1,49 +1,177 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Factorial
+
+# Time complexity: O(n)
+# Space complexity: O(n)
+
+# factorial(n) Write a method factorial that accepts an integer parameter n and that uses recursion to compute and return the value of n factorial (also known as n!).
+
+# e.g. fact(4) = 4 * 3 * 2 * 1 = 24
+
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError, "Please use a positive number"
+  elsif n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Reverse
+
+# Write a method `reverse` that accepts a string as a parameter and then returns the reverse of the string by reversing all letters and all words in the string.
+# - e.g. reverse("hello world") will return "dlrow olleh"
+
+# Time complexity: O(n)
+# Space complexity: 0(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+
+  reversed_s = reverse(s[1..-1])
+  reversed_s += s[0]
+  return reversed_s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Reverse in place
+
+# Write a method `reverse_inplace` that accepts a string as a parameter and then reverses the string in place using a recursive algorithm.
+# - e.g. reverse("hello world") will convert the input string to "dlrow olleh"
+
+# From example in class
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_help(s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_help(s, first = 0, last = s.length - 1)
+  if first < last
+    s[first], s[last] = s[last], s[first]
+    return reverse_help(s, first + 1, last - 1)
+  end
+
+  return s
+end
+
+# Bunny ears
+
+# Write a method `bunny` that accepts an integer parameter n. N represents a number of bunnies and each bunny has two big floppy ears. We want to compute the total number of ears across all the bunnies recursively (without loops or multiplication).
+# - e.g. bunny(0) = 0
+# - e.g. bunny(1) = 2
+# - e.g. bunny(10) = 20
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Nested
+
+# Write a method `nested` that accepts a string of only parenthesis and then returns if those parenthesis are properly nested. You may
+# assume that no non-parenthesis characters will be passed to this method.
+# - e.g. nested("((()))") = true
+# - e.g. nested("())") = false
+# **Challenge**: Try doing this without creating new strings in the process of solving the problem.
+
+# Time complexity: O(n)
+# Space complexity: o(n^2)
 def nested(s)
```

Comment:
You're not creating a new array with each recursive call so the space complexity is O(n)